bdk_core/tests/test_sparse_chain.rs

```diff
@@ -274,93 +222,74 @@ fn adding_checkpoint_where_tx_conftime_has_changed() {
-    let mut update2 = checkpoint_gen.create_update(
-        vec![TxSpec {
-            inputs: vec![ISpec::Other],
-            outputs: vec![OSpec::Mine(2_900, 0)],
-            confirmed_at: Some(0),
-        }],
-        0,
-    );
-
-    update2.base_tip = None;
 
+    let update2 = CheckpointCandidate {
+        base_tip: Some(BlockId {
+            height: 0,
+            hash: BlockHash::all_zeros(),
+        }),
+        ..checkpoint_gen.create_update(
+            &mut graph,
+            vec![TxSpec {
+                inputs: vec![ISpec::Other],
+                outputs: vec![OSpec::Mine(2_900, 0)],
+                confirmed_at: Some(2),
+            }],
+            2,
+        )
+    };
     assert_eq!(
         chain.apply_checkpoint(update2.clone()),
         ApplyResult::Stale(StaleReason::BaseTipNotMatching {
-            got: None,
-            expected: chain.latest_checkpoint().unwrap()
+            got: chain.latest_checkpoint(),
+            expected: update2.base_tip.unwrap(),
         })
     );
 
-    update2.invalidate = Some(update1.new_tip);
-
-    assert_eq!(chain.apply_checkpoint(update2.clone()), ApplyResult::Ok);
-    assert_eq!(chain.iter_tx().count(), 1);
-    assert_eq!(chain.iter_tx().next().unwrap().1.tx.output[0].value, 2_900);
-}
-
-#[test]
-fn simple_double_spend_inconsistency_check() {
-    let mut checkpoint_gen = CheckpointGen::new();
-    let mut chain = SparseChain::default();
-    let first = checkpoint_gen.create_update(
-        vec![TxSpec {
-            inputs: vec![ISpec::Other],
-            outputs: vec![OSpec::Mine(1_000, 0)],
-            confirmed_at: Some(0),
-        }],
-        0,
-    );
-
-    let txid = first.transactions[0].tx.txid();
-
-    let second = checkpoint_gen.create_update(
-        vec![TxSpec {
-            inputs: vec![ISpec::Explicit(OutPoint { txid, vout: 0 }), ISpec::Other],
-            outputs: vec![OSpec::Mine(2_000, 0)],
-            confirmed_at: Some(1),
-        }],
-        1,
-    );
-
-    let third = checkpoint_gen.create_update(
-        vec![TxSpec {
-            inputs: vec![ISpec::Explicit(OutPoint { txid, vout: 0 }), ISpec::Other],
-            outputs: vec![OSpec::Mine(2_000, 1)],
-            confirmed_at: Some(1),
-        }],
-        1,
-    );
-
-    assert_eq!(chain.apply_checkpoint(first), ApplyResult::Ok);
-    assert_eq!(chain.apply_checkpoint(second.clone()), ApplyResult::Ok);
-    assert_eq!(
-        chain.apply_checkpoint(third.clone()),
-        ApplyResult::Inconsistent {
-            txid: third.transactions[0].tx.txid(),
-            conflicts_with: second.transactions[0].tx.txid(),
-        }
-    );
+    let update3 = CheckpointCandidate {
+        base_tip: Some(BlockId {
+            height: 1,
+            hash: BlockHash::all_zeros(),
+        }),
+        invalidate: Some(update1.new_tip),
+        ..update2.clone()
+    };
+    assert_eq!(chain.apply_checkpoint(update3.clone()), ApplyResult::Ok);
```

Comment:
This checkpoint doesn't make sense. it's invalidate is the same as it's base_tip. Should be an error.

Reply:
I may have misunderstood either invalidate or base tip, but why does it need to be an error here?

Base tip is saying, we can only apply this update if the tip of sparsechain is this height and hash right?

And invalidate means, from this height and hash, all blocks and transactions should be removed?

Reply:
> I may have misunderstood either invalidate or base tip, but why does it need to be an error here?
> 
> Base tip is saying, we can only apply this update if the tip of sparsechain is this height and hash right?

No. The fields mean:

- `base_tip` :  The **latest** checkpoint that is in the chain.
- `invalidate`: The **earliest** checkpoint that is not in the chain.

It might help to rename the fields to reflect this better.

Reply:
Let me know if my latest changes make sense.